### ISSUE TITLE
Rule proposal: Forbid afterEach etc callback with serial

### ISSUE THREAD

**Issue by novemberborn** [avajs/eslint-plugin-ava#166]
This implies the callback only applies to serial tests, which is not the case. If I'm not mistaken regular `test.afterEach` hooks still run serially when run after a `test.serial()` test.

See <https://github.com/avajs/ava/issues/1178>.

### Fail

``` js
test.serial.before(fn);
test.serial.beforeEach(fn);
test.serial.afterEach(fn);
test.serial.after(fn);
test.serial.afterEach.always(fn);
test.serial.after.always(fn);
```
### Valid

``` js
test.before(fn);
test.beforeEach(fn);
test.afterEach(fn);
test.after(fn);
test.afterEach.always(fn);
test.after.always(fn);
```

**Comment by sindresorhus** [comment]
It does have a purpose though:

```js
// This would run first
test.serial.beforeEach(fn);

// Then these would run concurrently
test.beforeEach(fn);
test.beforeEach(fn);
```

**Comment by novemberborn** [comment]
No, hooks run in order of registration, and for serial tests they automatically run serially:

```js
import test from 'ava'

test.beforeEach(async () => {
  console.log('1.1', new Date())
  await new Promise(resolve => setTimeout(resolve, 1000))
  console.log('1.2', new Date())
})

test.serial.beforeEach(() => {
  console.log('2', new Date())
})

test.beforeEach(async () => {
  console.log('3.1', new Date())
  await new Promise(resolve => setTimeout(resolve, 1000))
  console.log('3.2', new Date())
})

test.beforeEach(async () => {
  console.log('4.1', new Date())
  await new Promise(resolve => setTimeout(resolve, 1000))
  console.log('4.2', new Date())
})

test.serial('5', t => console.log('5', new Date()))
test('6', t => console.log('6', new Date()))
test('7', t => console.log('7', new Date()))
```

Gives (note the timestamps):

```console
$ "$(npm bin)"/ava --verbose test.js

1.1 2017-01-12T16:25:58.964Z
1.2 2017-01-12T16:25:59.970Z
2 2017-01-12T16:25:59.975Z
3.1 2017-01-12T16:25:59.976Z
3.2 2017-01-12T16:26:00.981Z
4.1 2017-01-12T16:26:00.982Z
4.2 2017-01-12T16:26:01.986Z
5 2017-01-12T16:26:01.986Z
  ✔ 5
1.1 2017-01-12T16:26:01.988Z
1.1 2017-01-12T16:26:01.989Z
1.2 2017-01-12T16:26:02.994Z
1.2 2017-01-12T16:26:02.994Z
2 2017-01-12T16:26:02.995Z
2 2017-01-12T16:26:02.995Z
3.1 2017-01-12T16:26:02.995Z
3.1 2017-01-12T16:26:02.995Z
3.2 2017-01-12T16:26:04.000Z
3.2 2017-01-12T16:26:04.000Z
4.1 2017-01-12T16:26:04.001Z
4.1 2017-01-12T16:26:04.001Z
4.2 2017-01-12T16:26:05.003Z
4.2 2017-01-12T16:26:05.003Z
6 2017-01-12T16:26:05.003Z
  ✔ 6
  ✔ 7
7 2017-01-12T16:26:05.004Z

  3 tests passed [16:26:05]
```

**Comment by sindresorhus** [comment]
Oh, right. Then I think we should disallow that combination in AVA too, at runtime, with a user-friendly message.

**Comment by mightyiam** [comment]
No need for a rule for something that gets an error thrown, right?

**Comment by sindresorhus** [comment]
@mightyiam There is. It's useful with an early error that can be caught in the editor even before running the tests.

**Comment by novemberborn** [comment]
> Oh, right. Then I think we should disallow that combination in AVA too, at runtime, with a user-friendly message.

#1182.

**Comment by sindresorhus** [comment]
@novemberborn I just tried `test.serial.before(() => {});` in the latest AVA 1.0.0 beta and it doesn't throw.

**Comment by sindresorhus** [comment]
@novemberborn Should we close this issue in favor of https://github.com/avajs/eslint-plugin-ava/issues/186?

**Comment by novemberborn** [comment]
Yup. `test.serial.before()` is now a serially-running before hook, and works as expected.
